pennylane/hf/tapering.py

```diff
@@ -0,0 +1,204 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for tapering qubits using symmetries.
+"""
+from copy import deepcopy
+import numpy as np
+import pennylane as qml
+
+
+def _binary_matrix(terms, num_qubits):
+    """Get a binary matrix representation of the hamiltonian where each row coressponds to a
+    Pauli term, which is represented by concatenation of Z and X vectors.
+
+    Args:
+        terms (Iterable[Observable]): operators defining the Hamiltonian.
+        num_qubits (int): number of wires required to define the Hamiltonian.
+    Returns:
+        E (ndarray): binary matrix representation of the Hamiltonian of shape
+        :math:`len(terms) \times 2*num_qubits`.
+
+    .. code-block::
+        >>> terms = [PauliZ(wires=[0]) @ PauliX(wires=[1]), PauliZ(wires=[0]) @ PauliY(wires=[2]),
+                        PauliXwires=[0]) @ PauliY(wires=[3])]
+        >>> get_binary_matrix(terms, 4)
+         array([[1, 0, 0, 0, 0, 1, 0, 0],
+                [1, 0, 1, 0, 0, 0, 1, 0],
+                [0, 0, 0, 1, 1, 0, 0, 1]]))
+
+    """
+
+    E = np.zeros((len(terms), 2 * num_qubits), dtype=int)
+    for idx, term in enumerate(terms):
+        ops, wires = term.name, term.wires
+        if len(term.wires) == 1:
+            ops = [ops]
+        for op, wire in zip(ops, wires):
+            if op in ["PauliX", "PauliY"]:
+                E[idx][wire + num_qubits] = 1
+            if op in ["PauliZ", "PauliY"]:
+                E[idx][wire] = 1
+
+    return E
+
+
+def _reduced_row_echelon(binary_matrix):
+    """ Returns the reduced row echelon form (RREF) of a matrix in a binary finite field Z_2 """
+
+    binary_matrix = deepcopy(binary_matrix)
```

Comment:
For my own learning purposes: why do we create this deep copy of the matrix?

Reply:
A shallow copy just creates a variable which points to the same object in memory. So for example if I do: 
```
A = [0, 1, 2]
B = A 

B[0] = 3
 
>>> B 
[3, 1, 2]

>>> A
[3, 1, 2]
```

So when multiple functions need to use the array object, you use ``deep copy`` to make sure that it doesn't change the core object in between.